content/post/codesent.md

```diff
@@ -0,0 +1,396 @@
+discover input and output tensors. The new session is launched and we execute the resolved dependency
+tree to reach the output given the input. Sessions are required to allocate the hardware resources
+and manage the environment. Technically, they are bridges to the underlying Tensorflow runtime
+written in C++ as a standalone library, in case with Go this bridge is implemented with CGo.
+
+### BiDiSentiment
+
+{{% caption src="/post/codesent/gophers.png" %}}
+vmarkovtsev/BiDiSentiment logo.
+{{% /caption %}}
+
+What's left is to reveal the comment sentiment classification model. I named it "BiDiSentiment" and
+this is it's logo. BiDiSentiment is a general-purpose sentiment model, it is not specifically designed
+for comments.
+
+{{% caption src="/post/codesent/arch_black.png" %}}
+BiDiSentiment architecture.
+{{% /caption %}}
+
+The model is by far not the state-of-the-art but it works reasonably well. It consists of two
+stacked recurrent branches, each branch has two LSTM layers. Both read text byte by byte, the first
+from left to right and the second in the opposite direction. Once all the text is read, their
+outputs are concatenated and the dense layer put on top. The result is two numbers: the estimated
+probabilities of the negative and the positive sentiment. Obviously, their sum is 1.
+
+Since the recurrent branches are independent, we can pin them to different GPUs which speeds up
+the training by 50%.
+
+{{% caption src="/post/codesent/arch_device_black.png" %}}
+BiDiSentiment training device allocation.
+{{% /caption %}}
+
+The biggest labelled dataset that I found was 1.5 million tweets so I trained on it. As a bonus,
+the text sizes are bound to 140 (now 280) characters which perfectly suits the maximum recurrent memory
+depth. Here is the plot of the accuracy (number of correctly predicted sentiments divided by the overall
+size) through time.
+
+{{% caption src="/post/codesent/validation_black.png" %}}
+BiDiSentiment accuracy on validation.
+{{% /caption %}}
+
+We split the data into two parts: the part on which we train and the part on which we evaluate aka
+validation. This is needed to avoid overfitting - when the model adapts to the training examples
+too much it loses generalization ability. The validation accuracy is plotted here. We see that
+our model overfits after 5 epochs - that is, runs across the whole training set. This is a typical
+situation in recurrent neural networks. I should have used overfit reduction techniques,
+e.g. dropout, but I did not have time. So we pick the graph state on epoch 5 and export it.
+
+Let's see how BiDiSentiment performs on some answers to that StackOverflow questions.
+
+```
+go get -v gopkg.in/vmarkovtsev/BiDiSentiment.v1/...
+
+echo "When I wrote this, only God and I understood \
+what I was doing. Now, God only knows" | $GOPATH/bin/sentiment
+```
+
+Output (negative probability): 0.8803515 - the sentiment is clearly negative.
```

Comment:
Could you also give how the sample was originally labeled?

Reply:
This is https://stackoverflow.com/a/316233 there is not label :)